# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [11]:
import pandas as pd
import re
df=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')

In [12]:
df2=df.copy()

In [13]:
df

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [14]:
df['Response'].map(str)

0         No
1         No
2         No
3        Yes
4         No
        ... 
10905     No
10906     No
10907     No
10908     No
10909    nan
Name: Response, Length: 10910, dtype: object

In [15]:
df.columns

Index(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'],
      dtype='object')

In [16]:
df=df[(df['Total Claim Amount']>1000) & (df['Response'] == 'Yes')].reset_index()

In [17]:
df

index  Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0     189         189  OK31456  California             11009.130490      Yes   
1     236         236  YJ16163      Oregon             11009.130490      Yes   
2     419         419  GW43195      Oregon             25807.063000      Yes   
3     442         442  IP94270     Arizona             13736.132500      Yes   
4     587         587  FJ28407  California              5619.689084      Yes   
..    ...         ...      ...         ...                      ...      ...   
62  10351       10351  FN44127      Oregon              3508.569533      Yes   
63  10373       10373  XZ64172      Oregon             10963.957230      Yes   
64  10487       10487  IX60941      Oregon              3508.569533      Yes   
65  10565       10565  QO62792      Oregon              7840.165778      Yes   
66  10708       10708  CK39096      Oregon              5619.689084      Yes   

    Coverage             Education Effective To Date EmploymentStatus  ...  \
0    Premium              Bachelor           1/24/11         Employed  ...   
1    Premium              Bachelor           1/24/11         Employed  ...   
2   Extended               College           2/13/11         Employed  ...   
3    Premium                Master           2/13/11         Disabled  ...   
4    Premium  High School or Below           1/26/11       Unemployed  ...   
..       ...                   ...               ...              ...  ...   
62  Extended               College            1/5/11    Medical Leave  ...   
63   Premium  High School or Below            2/8/11         Employed  ...   
64  Extended               College            1/5/11    Medical Leave  ...   
65  Extended               College           1/14/11         Employed  ...   
66   Premium  High School or Below           1/26/11       Unemployed  ...   

   Number of Open Complaints  Number of Policies     Policy Type  \
0                        0.0                   1  Corporate Auto   
1                        0.0                   1    Special Auto   
2                        1.0                   2   Personal Auto   
3                        0.0                   8   Personal Auto   
4                        0.0                   1   Personal Auto   
..                       ...                 ...             ...   
62                       1.0                   1   Personal Auto   
63                       0.0                   1  Corporate Auto   
64                       1.0                   1   Personal Auto   
65                       2.0                   1   Personal Auto   
66                       0.0                   1   Personal Auto   

          Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0   Corporate L3            Offer2          Agent         1358.400000   
1     Special L3            Offer2          Agent         1358.400000   
2    Personal L2            Offer1         Branch         1027.200000   
3    Personal L2            Offer1            Web         1261.319869   
4    Personal L1            Offer2            Web         1027.000029   
..           ...               ...            ...                 ...   
62   Personal L2            Offer2         Branch         1176.278800   
63  Corporate L2            Offer1          Agent         1324.800000   
64   Personal L3            Offer2         Branch         1176.278800   
65   Personal L3            Offer2          Agent         1008.000000   
66   Personal L3            Offer2            Web         1027.000029   

    Vehicle Class  Vehicle Size Vehicle Type  
0      Luxury Car       Medsize          NaN  
1      Luxury Car       Medsize            A  
2      Luxury Car         Small            A  
3             SUV       Medsize            A  
4             SUV       Medsize            A  
..            ...           ...          ...  
62  Four-Door Car         Small          NaN  
63     Luxury SUV       Medsize     

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [18]:
df2=df2[df2['Response'] == 'Yes']
print(df2['Monthly Premium Auto'].mean())
df_Clv_PolicyType_Gender=df2.pivot_table(index=['Policy Type', 'Gender'], values=['Total Claim Amount', 'Customer Lifetime Value', 'Monthly Premium Auto'], aggfunc='mean')
#calculating
df_Clv_PolicyType_Gender['Cost/Expected Revenue']=((df_Clv_PolicyType_Gender['Total Claim Amount'])/(df_Clv_PolicyType_Gender['Customer Lifetime Value'])*100)
df_Clv_PolicyType_Gender['Breakeven']=((df_Clv_PolicyType_Gender['Total Claim Amount'])/(df_Clv_PolicyType_Gender['Monthly Premium Auto']))
#formating
df_Clv_PolicyType_Gender['Monthly Premium Auto']=df_Clv_PolicyType_Gender['Monthly Premium Auto'].round(2)
df_Clv_PolicyType_Gender['Total Claim Amount']=df_Clv_PolicyType_Gender['Total Claim Amount'].round(2)
df_Clv_PolicyType_Gender['Customer Lifetime Value']=df_Clv_PolicyType_Gender['Customer Lifetime Value'].round(2)
df_Clv_PolicyType_Gender['Cost/Expected Revenue']=df_Clv_PolicyType_Gender['Cost/Expected Revenue'].round(2).apply(lambda x: f'{x}%')
df_Clv_PolicyType_Gender['Breakeven']=df_Clv_PolicyType_Gender['Breakeven'].round(2)
df_Clv_PolicyType_Gender

94.41268758526603


Customer Lifetime Value  Monthly Premium Auto  \
Policy Type    Gender                                                  
Corporate Auto F                       7712.63                 94.30   
               M                       7944.47                 92.19   
Personal Auto  F                       8339.79                 99.00   
               M                       7448.38                 91.09   
Special Auto   F                       7691.58                 92.31   
               M                       8247.09                 86.34   

                       Total Claim Amount Cost/Expected Revenue  Breakeven  
Policy Type    Gender                                                       
Corporate Auto F                   433.74                 5.62%       4.60  
               M                   408.58                 5.14%       4.43  
Personal Auto  F                   452.97                 5.43%       4.58  
               M                   457.01                 6.14%       5.02  
Special Auto   F                   453.28                 5.89%       4.91  
               M                   429.53                 5.21%       4.97

The most profitable and less risk for the company are:
- Corporate policy, female, with an average of 6.33% of costs for the expected revenue; and Personal policy, female, with an av. of 8.52% of costs for the expected revenue.
- Women customers are, in average, less risky.
- Nevertheless, the total claims are equivalent to aproximately 4.8 months of premiums regardless the type and gender.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [19]:
st=df2['State'].value_counts()

Only California has more than 500
State
California    504
Oregon        419
Arizona       281
Nevada        140
Washington    122

In [20]:
# 1. Calculate the counts for all states.
#state_counts = df2['State'].value_counts()

# 2. Identify the index (the state names) of the states that have counts > 500.
# The result is a list-like Index object containing only the high-frequency state names.
#StatesToKeep = state_counts[state_counts > 500].index

# 3. Filter the entire DataFrame df2 using .isin() and assign the result back to df2.
#df2 = df2[df2['State'].isin(StatesToKeep)]

# Display the resulting filtered DataFrame
#df2

count=df2['State'].value_counts()
StatesToKeep=count[count>500].index
df2=df2[df2['State'].isin(StatesToKeep)]
df2

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
8               8  FM55990  California              5989.773931      Yes   
15             15  CW49887  California              4626.801093      Yes   
19             19  NJ54277  California              3746.751625      Yes   
122           122  EW33419  California             15114.402410      Yes   
125           125  BM92346  California              3919.366722      Yes   
...           ...      ...         ...                      ...      ...   
10750       10750  AY52869  California              4386.277558      Yes   
10757       10757  TK45392  California              5107.163002      Yes   
10780       10780  JW94491  California             11790.496170      Yes   
10823       10823  AH53588  California              4636.546470      Yes   
10872       10872  XT67997  California              5979.724161      Yes   

       Coverage             Education Effective To Date EmploymentStatus  \
8       Premium               College           1/19/11         Employed   
15        Basic                Master           1/16/11         Employed   
19     Extended               College           2/26/11         Employed   
122       Basic  High School or Below           1/19/11    Medical Leave   
125    Extended  High School or Below            1/5/11         Employed   
...         ...                   ...               ...              ...   
10750     Basic               College           2/11/11          Retired   
10757     Basic              Bachelor           1/26/11         Employed   
10780     Basic               College            1/3/11          Retired   
10823     Basic               College            1/1/11          Retired   
10872  Extended  High School or Below            2/7/11         Employed   

      Gender  ...  Number of Open Complaints Number of Policies  \
8          M  ...                        0.0                  1   
15         F  ...                        0.0                  1   
19         F  ...                        1.0                  1   
122        F  ...                        1.0                  2   
125        F  ...                        0.0                  1   
...      ...  ...                        ...                ...   
10750      M  ...                        0.0                  5   
10757      F  ...                        1.0                  3   
10780      M  ...                        0.0                  2   
10823      M  ...                        1.0                  5   
10872      F  ...                        0.0                  3   

          Policy Type        Policy  Renew Offer Type  Sales Channel  \
8       Personal Auto   Personal L1            Offer2         Branch   
15       Special Auto    Special L1            Offer2         Branch   
19      Personal Auto   Personal L2            Offer2    Call Center   
122    Corporate Auto  Corporate L2            Offer2    Call Center   
125     Personal Auto   Personal L1            Offer3          Agent   
...               ...           ...               ...            ...   
10750   Personal Auto   Personal L1            Offer1          Agent   
10757   Personal Auto   Personal L2            Offer2          Agent   
10780   Personal Auto   Personal L2            Offer2            Web   
10823   Personal Auto   Personal L2            Offer2          Agent   
10872   Personal Auto   Personal L3            Offer2          Agent   

       Total Claim Amount  Vehicle Class Vehicle Size Vehicle Type  
8              739.200000     Sports Car      Medsize          NaN  
15             547.200000            SUV      Medsize          NaN  
19              19.575683   Two-Door Car        Large            A  
122            480.000000            SUV      Medsize            A  
125            558.099357  Four-Door Car        Small          NaN  
...                   ...            ...          ...          ...  
10750          321.600000  F

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [25]:
CLdf=df2.pivot_table(index=['Education', 'Gender'], values='Customer Lifetime Value', aggfunc= ['max', 'min', 'median'])
CLdf=CLdf.round(2)
CLdf

max                     min  \
                            Customer Lifetime Value Customer Lifetime Value   
Education            Gender                                                   
Bachelor             F                     33473.35                 2248.45   
                     M                     24127.50                 2300.69   
College              F                     25807.06                 2004.35   
                     M                     22332.44                 2393.92   
Doctor               F                      6265.34                 2395.57   
                     M                     12731.95                 3436.13   
High School or Below F                     41787.90                 2397.04   
                     M                     12298.69                 2227.07   
Master               F                     19160.99                 2574.02   
                     M                     14435.67                 2551.23   

                                             median  
                            Customer Lifetime Value  
Education            Gender                          
Bachelor             F                      5217.51  
                     M                      4984.62  
College              F                      6273.91  
                     M                      6005.85  
Doctor               F                      4014.45  
                     M                     10972.07  
High School or Below F                      5979.72  
                     M                      6366.23  
Master               F                      4861.74  
                     M                      8509.85

In [ ]:
- Regarding the minimum, differences are negligible, both populations behave similarly.
- The maximum show much higher value in women compared with men with the same education, apart from the 'Doctors', where men retrieve almost the doble. 
- Although the median is in accordance with max. and min. for Doctors (and likely with the average), the median of people with Master education has a contrast with the maxs and mins, showing that there is at least 1 woman outstanding. But men with Master population represent more value.

The conclusion: maximum and minimum don't represent population, but median does. Men with Doctor and Master represent higher value than women, and the other levels of education they don't have a significant difference.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [22]:
# your code goes here